### ISSUE TITLE
add include patterns

### ISSUE THREAD

**Issue by leo-b** [borgbackup/borg#1406]
Sometimes the ability to not only exclude patterns but also explicitly specify includes can be useful.

I propose a new optional pattern prefix '+' that may be used with --exclude or in --exclude-from files to define include-patterns. (Just like include/exclude pattern rules in rsync.)

These patterns may be used to explicitly include files in a backup that would otherwise be excluded by a following pattern.

Some use cases:
- exclude docs but backup pdfs:
  
  ```
  --exclude +/data/docs/pdf --exclude /data/docs/
  ```
  
  (maybe `--include /data/docs/pdf` should be introduced to improve readability.)
- use a common exclude list deployed to all backup clients but override it on some specific client:
  
  `global.inclexcl`:
  
  ```
  /tmp
  *.tmp
  /scratch
  ```
  
  `local.inclexcl`:
  
  ```
  +/scratch
  /home/dir_to_exclude
  ```
  
  `$ borg create --exclude-from local.inclexcl --exclude-from global.inclexl backup /`
  
  equivalent:
  
  `global.inclexcl`:
  
  ```
  -sh:/tmp
  -fm:*.tmp
  -fm:/scratch
  ```
  
  `local.inclexcl`:
  
  ```
  +fm:/scratch
  -/home/dir_to_exclude
  ```

(Actually the same approach is already used by the --keep-tag-files option.)

**Comment by PlasmaPower** [comment]
Filenames can start with + though.

**Comment by leo-b** [comment]
> Filenames can start with + though.

Use ``-+file`` or ``pm:+file`` to escape and exclude a file named ``+file``.
It's just like you have to escape files named ``xx:file``:
> Explicit style selection is necessary when a non-default style is desired or when the desired pattern starts with two alphanumeric characters followed by a colon (i.e. aa:something/*).

**Comment by PlasmaPower** [comment]
@leo-b It still would be a breaking change, although probably nobody's excluding a file starting with `+`.

**Comment by leo-b** [comment]
@PlasmaPower I agree it should be noted in the release notes for academic purposes... ;-)

**Comment by ThomasWaldmann** [comment]
@leo-b don't you find giving include patterns with --exclude **weird** somehow?

**Comment by enkore** [comment]
That weirdness is shared with some other programs as well (`.*ignore` files are notorious here).

I'm not sure if this is really needed. It feels kinda rare use case and since both exclude and paths support regexpes the functionality is already there (although more complex to use).

**Comment by leo-b** [comment]
@ThomasWaldmann as suggested: we could introduce an ``--include`` option to make it more intuitively.

Of course you can use regex negative lookahead assertions to except files from exclusion but those regexes will become **really** complicated and unreadable.

Besides my second usecase (deploying a global default exlude-list with the ability to override it with a per-client filter-file) wouldn't be possible with existing functionality.

**Comment by leo-b** [comment]
Other tools with include/exclude syntax:
- rsync: https://download.samba.org/pub/rsync/rsync.html
  (see INCLUDE/EXCLUDE PATTERN RULES)

- rdiff-backup: http://www.nongnu.org/rdiff-backup/examples.html#exclude

- duplicity: http://duplicity.nongnu.org/duplicity.1.html#sect9

- IBM Tivoli Storage Client: (patterns are evaluates from bottom up here)
  http://www.ibm.com/support/knowledgecenter/SSTG2D_7.1.0/com.ibm.itsm.client.doc/t_cfg_crtinclexcl.html

... and probably many more. :-)

**Comment by ovizii** [comment]
Just curios if I can include certain file types right now, aka run borg like this: 

`/usr/local/sbin/borg create -s -C zlib,6 -v  $REPOSITORY::`hostname`-`date +%Y-%m-%d`   /var/lib/vz/dump/vzdump-lxc-102-*`

I'm interesting if this can work: 
**vzdump-lxc-102-***

**Comment by enkore** [comment]
Unquoted stars and other glob patterns like in your example are interpreted by the shell

**Comment by ovizii** [comment]
@enkore thanks. is there some more explanation somewhere? couldn't find anything in the docs: https://borgbackup.readthedocs.io but maybe I am searching for the wrong terms?

**Comment by enkore** [comment]
These are documented in the shell manual e.g.

http://wiki.bash-hackers.org/syntax/pattern
https://www.gnu.org/software/bash/manual/html_node/Pattern-Matching.html
http://zsh.sourceforge.net/Intro/intro_2.html

**Comment by piotrturski** [comment]
i also vote for this feature. another possible syntax is not to make `-e` more complicated but add another one `-i` (or similar) with same patterns semantic as `-e` and let them be written in order:
```
-e ~/docs        # exclude docs \
-i ~/docs/pdf      # but retain pdfs \
-e ~/docs/pdf/useless-pdf    # but exclude useless pdfs
```

**Comment by piotrturski** [comment]
another possibility would be to allow delegation of file selection to external program. like taking them from pipe or doing something similar to `find -exec`

**Comment by ThomasWaldmann** [comment]
I think having include / exclude rules would be nice to have (we somehow have them in a limited way via borg create repo:arch path1 path2... (pathX is a include) and --exclude(-from), but everything could be prettier and more regular somehow).

What I'ld find awkward is to have include rules via --exclude or --exclude-from.
Or patterns in such a file being exclude by default and include if prefixed by "+".

So, to unblock this, we need to make up something that doesn't have these weirdnesses:

- do not change semantics of current --exclude(-from)
- introduce 2 new options `--xxx` and `--xxx-from` (find some reasonable word for xxx) that refers to patterns that can be include or exclude (or maybe even other things by using more prefixes).
- `--xxx -foo` would mean exclude foo, `--xxx +foo` would mean include foo
- `--xxx-from FILE` would contain lines like:

```
# comment
+ includepattern
- excludepattern
X anything else than + and - can be used for future extension
```

**Comment by piotrturski** [comment]
then my spec proposition is:

`--filter` and `--filter-from`. every entry has to be prefixed with `+` or `-`, then pattern (in same format as for exclusions), that is:
```
--filter +~/Documents \
--filter -'re:/home/user/*.log'
```
when used together with `--exclude`, `--filter` is treated as added 'later', that means, has higher priority, therefore can be used to fine-tune all currently existing commands with `--exclude` without a need to rewrite them completely

when read from file, space is allowed between `+`/`-` and the pattern (is it possible also with command line?):
```
+ ~/Documents
- 're:/home/user/*.log'
```
(i never used exclusions from file so i don't know if quoting rules are the same)

order of patterns is important. later patterns are more important (have higher priority) and override the previous ones (lower priority). when combining different ways of providing pattern, the order is as follows:

- `--exclude-from` and ` --exclude` (lowest priority)
- `--filter-from`
- `--filter` (highest priority; will override previous filters and exclusions)

`-e` and `--exclude` become deprecated

**Comment by ThomasWaldmann** [comment]
Another thing that comes to mind: maybe `borg create` should be able to work without `pathN` commandline args if just everything is given in that `--xxx-from file`?

**Comment by piotrturski** [comment]
may be a bit misleading. what should be backed up if there is only `- /tmp`? all except tmp? so we implicitly add `+ /`. but what if there is only `+ ~/Documents`? should we also implicitly add `+ /`? of course not. so we require to have `+ ...` as a first param? but what if we get `+ 're:(!(/tmp|/proc))'`? should we also implicitly add `+ /`?

i think some clear starting points are needed. i'm not saying it won't work without explicit `path`, i'm just saying it seems to be much more complicated to design it in a way that is clear for users.

**Comment by ThomasWaldmann** [comment]
It's not misleading, it is just a logical conclusion if one goes away from that file being an *exclude* file (not sure if *filter* is much better), but also having includes now.

In general, borg does not hold you back from doing stuff that does not make sense, so let's not consider cases that just do not make sense.

So, assume that we start from nothing. Nothing included, nothing excluded, no paths on borg commandline:
```
borg create repo::arch --xxx-from filename
```
filename contents:
```
+ /
- /dev
- /proc
- /sys
- /home            # we hate all users ;)
+ /home/love1      # except that one
+ /home/love2      # except that one
- /home/*/.cache   # can be rebuild
...
```
I find this quite straightforward. Also, one has all in one place.

**Comment by ThomasWaldmann** [comment]
As each line is a pattern, maybe `--patterns-from FILE` and `--pattern PAT`?

**Comment by piotrturski** [comment]
you're right, starting from nothing is essentially adding `- /` at the beginning and is pretty clear. 

i don't know how it's implemented but what will borg do if it gets `+ 're:...'`? will it scan entire folder structure starting from `/`? including `/proc`, `/run` etc? is it efficient enough? if it's not a problem, then yeah, seems like the spec is complete

**Comment by ThomasWaldmann** [comment]
Well, we do not have include patterns yet, so that never was a problem yet.

It does not only happen for the initial include pattern, it's logically the same if you need to consider such includes "below" excluded paths.

**Comment by piotrturski** [comment]
so i considered `path` arguments as starting point. i assume currently only `path` directories are scanned and matched against exclude patterns. i was thinking about keeping same rule with inclusions. that is, if we have:
```
borg create ... \
    /a \
    /b \
   --pattern +/c
```
then folder `/c` would NOT be included because it's not inside any of the starting points (`/a` and `/b`). this actually makes complexity linear to what user wants to back up. otherwise you're right: full scan would be always needed

**Comment by ThomasWaldmann** [comment]
Maybe we could compute the starting points ("roots") from the given patterns (not from regexes ofc).

**Comment by piotrturski** [comment]
i think every time we encounter `+ 're:...'`we have to take `/` as root. simple case:
```
+ ~/Documents
- ~/Documents/pdf
+ 're:/home/user/Documents/pdf/(a|b)'
```
borg will have to analyse regex and expand shell variables like `~` to know that root is `~/Documents`. looks like huuuge task. we can add third marker `r` in addition to `+` and `-` but that's essentially same as `path`, just in different place

**Comment by ThomasWaldmann** [comment]
How about (when searching for roots):
- just consider "+" patterns
- remove all re: patterns
- remove every pattern that is prefixed by some other pattern
- what remains are the roots

**Comment by piotrturski** [comment]
won't work because:
```
+ 're:/logs-2016-??-??'
```
but you can tell users that only patterns like `+ simple-path` will be considered roots. for me it would be bad UX because it makes some functionality implicit. i would advice to require users to provide roots explicitly. in any form you prefer (param `path` or pattern `r`) but explicit

**Comment by piotrturski** [comment]
or you can add option to provide roots explicitly and infer roots if they are not provided using above flawed in general but probably working in most practical cases

**Comment by leo-b** [comment]
I like the naming  `--pattern` and `--patterns-from` .

### Concerning the file selection algorithm:
Most tools (rsync, rdiff-backup, duplicity) use this algorithm:
- the first matching pattern wins
- if the first matching pattern is an include pattern then backup this file. if it is an exclude pattern, exclude it. (And don't traverse deeper if the excluded object was a directory.)

I think we should adopt this logic, since users are accustomed to it.

So your (@ThomasWaldmann ) example:

> ```
> + /
> [...]
> - /home            # we hate all users ;)
> + /home/love1      # except that one
> + /home/love2      # except that one
> - /home/*/.cache   # can be rebuild
> ...
> ```

... would look like that:
```
+ /
[...]
- /home/*/.cache   # can be rebuild ...
+ /home/love1      # include that directory
+ /home/love1/*   # ... and its contents
+ /home/love2      # include that directory
+ /home/love2/*   # ... and its contents
- /home/*             # but exclude all other directorys under /home (not /home itself!)
```

### Concerning the selection of roots:
I think trying to detect if a pattern is a prefix-pattern of another could be rather complicated and error-prone. So in my opinion an explicit marker that defines roots would be preferable.
As an alternative a simple rule that would be easy to implement and to understand by users could be: Use every include pattern up to the first exclude pattern as roots.

**Comment by piotrturski** [comment]
@leo-b if first pattern matches then you will back up your entire system because you have `+ /` as a first pattern

**Comment by leo-b** [comment]
@piotrturski You are right. The implicit appending of a `*` is to blame for that:
Because `+ / ` ends with a slash, it is [internally translated](https://github.com/borgbackup/borg/blob/master/src/borg/helpers.py#L434) to `+ fm:/*`.
Unfortunately this makes it impossible to add only the root directory (without its subdirs) using the _fm:_ syntax. (`+ /.` would probably work but this is rather unintuitive.)
Maybe the implicit addition of the star should be eliminated?

Another reason why selecting the root should not be done implicitly via include patterns. ;-)

**Comment by enkore** [comment]
`# bla` -> comment

`r[oot] path` -> root def

`+ pattern`, `i[nclude] pattern`

`- pattern`, `e[xclude] pattern`

Empty lines ain't do nothing.

**Comment by piotrturski** [comment]
looks good. two things should be clarified: 

1. roots needs to be a simple path (no regex)
1. root's are read before all other patterns or in place/order where they are found?

**Comment by piotrturski** [comment]
suggestion: change
> By default all files are considered

to:

"By default all files under roots are considered"

And what will happen if there is multiple `root` lines? will it be error? will we sum them? will patterns under one `root` be separated from patterns under another?

**Comment by leo-b** [comment]
@enkore Your `root` proposal looks good to me. :-)

However, I think your example needs two small refinements..
> ```
> root /
> [...]
> # can be rebuild
> - /home/*/.cache
> # they're downloads for a reason
> - /home/*/Downloads
> # susan is a nice person
> + /home/susan
> # he's a cat, so backup his files
> + /home/richard
> # all others can go to hell
> - /home
> ```

The last line `- /home` would match and exclude the home directory itself, so borg wouldn't even traverse into the directory. Thus the upper lines won't ever be used. You could use `- /home/` (or equivalently `- /home/*`) instead.
Besides, `+ /home/susan` would include the directory but not its contents. So an additional `+ /home/susan/` would be necessary, otherwise `- /home/*` would match inside susans home.

Thus I'd suggest:
```
root /
[...]
# can be rebuild
- /home/*/.cache
# they're downloads for a reason
- /home/*/Downloads
# susan is a nice person
+ /home/susan
+ /home/susan/*
# he's a cat, so backup his files
+ /home/richard
+ /home/richard/*
# all others can go to hell
- /home/*
```

**Comment by enkore** [comment]
>    By default all files are considered
>
> to:
>
> "By default all files under roots are considered"

Kind of implied, but yes.

> And what will happen if there is multiple root lines? will it be error? will we sum them? will patterns under one root be separated from patterns under another?

I'd say that only patterns declared after a `root` occurred are considered for a root.

Whether in this:

```
root a
- pat1
- pat2
root b
- pat3
- pat4
```

`pat3`, `pat4` also apply to root a can be debated either way.

> The last line - /home would match and exclude the home directory itself, so borg wouldn't even traverse into the directory. Thus the upper lines won't ever be used.

Yep, has to be '- /home/*` or similar.

> Besides, + /home/susan would include the directory but not its contents. So an additional + /home/susan/ would be necessary, otherwise - /home/* would match inside susans home.

In fnmatch there shouldn't be a difference between these two variants. As soon as a directory is included it's contents are recursively included as well as long as no negative qualifier (--one-file-system, excludes) apply.

**Comment by leo-b** [comment]
>>   Besides, + /home/susan would include the directory but not its contents. So an additional + /home/susan/ would be necessary, otherwise - /home/* would match inside susans home.
> 
> In fnmatch there shouldn't be a difference between these two variants. As soon as a directory is included it's contents are recursively included as well as long as no negative qualifier (--one-file-system, excludes) apply.

Yep. But `- /home/*` would match its contents, so it would be excluded.

**Comment by enkore** [comment]
Why would `/home/*` match anything like `/home/xyz/1234`?

**Comment by leo-b** [comment]
Because when using _fnmatch_ (which is the default selector), `*` matches a path separator.

**Comment by enkore** [comment]
Oh well. That's kinda stupid. Dangerous, even. (`--exclude /something/*` and `--exclude '/something/*'` mean wildly different things)

So we'd have to do

```
...
- sh:/home/*
```

**Comment by enkore** [comment]
So I'd say that the default for these pattern-file-thingies should be `sh`, not `fnmatch`, because fnmatch is a stupid default for anything new.

**Comment by leo-b** [comment]
Agreed.
(Though it's confusing to have different defaults for `--exclude(-from)` and `--pattern(s-from)`.)

**Comment by enkore** [comment]
Yes. I guess fnmatch was modelled to behave like gitignore, which is somewhat similar, allowing things like `*.crap` matching globally.

**Comment by piotrturski** [comment]
another option would be to just drop backward compatibility and design it in a way that doesn't confuses anyone. then we would be able to just take existing filtering scheme from other backup tools. i just don't know if it's an acceptable solution

**Comment by enkore** [comment]
Hm.

Not sure if it's a good idea, but how 'bout making the default behave like:

- no / in pattern -> fnmatch semantics
- / in pattern -> sh semantics

I don't really see anyone writing `/home/*/junk` and expecting that `/home/user/subdir/junk` matches that.

**Comment by piotrturski** [comment]
i would advise to do as little implicit actions as possible. if someone is doing real backup, he creates a script for it. it's really no problem to for a user to add sh: or fn: It's much less error prone then changing wildcard meaning based on part of the actual pattern

**Comment by enkore** [comment]
That behaviour is relatively similar to rsync's pattern matching (not sure if that's a good thing -- [the rules are fairly complex](https://download.samba.org/pub/rsync/rsync.html)), which generally seems to "do what you expect".

The whole mechanism in rsync is pretty similar to what we are discussing here. Ie it has the "roots" specified from the command line, by default everything would be considered, and the include/exclude patterns refine on that.

**Comment by piotrturski** [comment]
> Risk: maybe rsync adds some stuff in the future (likelyhood?).

it's not a risk. borg won't use rsync's lib to do parsing/matching, right? so even when they change their syntax, it won't afect borg in any way. if you write in a doc that borg uses same syntax as rsync, just add rsync version and don't name the prefix `rsync`

>   this is completely different from anything in Borg, and doesn't really work with the existing xyz: stuff. OTOH could call this kind of pattern `rsync:`

exactly. make it another prefix. but remember this is completely separate task from `exclude`. i suggest starting from the `exclude` because it gives users much more powerful tool. and improving selectors' expressiveness maybe be a long iterative process (with users' feedback)

**Comment by leo-b** [comment]
I'd vote for keeping it rather simple and implement only rsyncs include/exclude pattern rules without modifiers and without any merge-filter rules.

IMHO the default (at least for the new `--pattern` and `--patterns-from` options) should be `sh:` syntax. If possible, I'd change the default for the other patterns too, for consistency.

**Comment by leo-b** [comment]
I've added a new pull request #1971  based on our discussion:

- Two new options `--pattern` and `--patterns-from` instead of modifying the behavior of `--exclude` and `--exclude-from`

- The new options default to shell-style (`sh:`) patterns. (Much cleaner inclusion of subdirs inside excluded dirs. Besides this leads to a behavior similar to rsync include/exclude patterns.)

- allow configurations of additional roots with `--pattern` and `--patterns-from` using a syntax like:
```
R /path
```

**Comment by ThomasWaldmann** [comment]
guess this should be closed?

**Comment by leo-b** [comment]
Closed, since follow up PR #1971 was already merged.
